design-proposals/klip-30-lambda-functions.md

```diff
@@ -0,0 +1,121 @@
+
+* `x => x + 1`
+* `s => UCASE(s)`
+* `(x, y) => x + y`
+* `(x, y) => CASE WHEN x IS NULL THEN y ELSE x + y END`
+* `(x, y) => x + COALESCE(y, 0)`
+
+### Invocation functions
+
+Lambda functions should require a specific **invocation function**, which tells ksqlDB how to apply the given lambda function to an input value, and what kind of result to return. The following invocation functions are proposed for the initial support for lambda functions:
+
+* `transform_array(arr, x => y)` - Applies the given lambda function to each element of the input `Array`, returning a new `Array` containing the transformed output.
+
+* `transform_map(map, (k, v) => x, y)` - Applies the given lambda function to each key-value pair of the input `Map`, returning a new `Map` containing the transformed key-value pairs.
+
+* `filter_array(arr, x => bool)` - Filters the input `Array` using the given lambda function. A new `Array` is returned, containing only values for which the lambda function evaluated to `true`.
+
+* `filter_map(map, (k, v) => bool)` - Filters the input `Map` using the given lambda function. A new `Map` is returned, containing only the key-value pairs for which the lambda function evaluated to `true`.
+
+* `reduce_array(arr, s, (x, s) => s)` - Reduces the input `Array` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_array` will return the final value of `s`.
+
+* `reduce_map(map, s, (k, v, s) => s)` - Reduces the input `Map` down to a single value. `s` is the initial state and is passed into the scope of the lambda function. Each invocation returns a new value for `s`, which the next invocation will receive. `reduce_map` will return the final value of `s`.
+
+## Documentation updates
+
+We must document and provide examples for the two core aspects of lambda functions:
+
+1. Lambda function syntax
+2. Invocation functions
+
+## Compatibility implications
+
+Lambda functions would be additive to ksqlDB's grammar and should therefore not introduce any incompatibilities into the grammar or public APIs. However, it is not currently known if a lambda function implementation would require backward-incompatible changes to ksqlDB's internals.
+
+## Security implications
+
+Since lambda functions would only allow SQL expressions that can already be used within queries, no new security implications should be introduced.
+
+## Open questions
+
+**1. Would it be possible to allow aggregates in lambdas within the context of a particular invocation function? For example,**
+
+- `agg_array(arr, x => sum(x))`
+
+If this is feasible then it would likely be worth doing. Without support for aggregates, users will not be able to easily perform any aggregation over a collection using `reduce_*`.
+
+**2. Would allowing lambda functions to be applied to types other than `Array` or `Map` provide any value?**
```

Comment:
I guess lambdas make sense for any "collection" data type. For scalar types, they seem useless because instead of writing `(x,y) => x + y` and apply it to two integers, one can just use the expression `x + y` of the lambda directly.

Reply:
My thoughts exactly but I wasn’t sure if I was missing something.

Reply:
I don't think it works for `STRUCT` given the disparate field types.  I don't think SQL can express how to handle the different types.